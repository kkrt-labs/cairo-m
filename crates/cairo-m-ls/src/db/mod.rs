//! Language Server database definitions and implementations.
//!
//! This module defines the concrete database used by the language server,
//! implementing all necessary traits from the compiler phases.

mod swapper;

use std::collections::HashMap;
use std::path::PathBuf;

use cairo_m_compiler_parser::{Db as ParserDb, SourceFile, Upcast};
use cairo_m_compiler_semantic::SemanticDb;
pub use swapper::AnalysisDatabaseSwapper;

/// The unified project crate representation used by the language server.
///
/// This serves as the single source of truth for project structure,
/// replacing the various ad-hoc crate creations throughout the codebase.
#[salsa::input(debug)]
pub struct ProjectCrate {
    /// The root directory of the project
    #[return_ref]
    pub root_dir: PathBuf,

    /// The main module name (e.g., "main" for main.cm)
    #[return_ref]
    pub main_module: String,

    /// All source files in the project, keyed by their absolute path
    #[return_ref]
    pub files: HashMap<PathBuf, SourceFile>,
}

/// The language server's analysis database.
///
/// This database extends the compiler database with additional
/// functionality needed for language server operations.
#[salsa::db]
#[derive(Clone, Default)]
pub struct AnalysisDatabase {
    storage: salsa::Storage<Self>,
}

impl AnalysisDatabase {
    /// Create a new analysis database.
    pub fn new() -> Self {
        Self::default()
    }
}

// Implement all required database traits
impl salsa::Database for AnalysisDatabase {}

#[salsa::db]
impl ParserDb for AnalysisDatabase {}

#[salsa::db]
impl SemanticDb for AnalysisDatabase {}

// Implement upcast traits for each database level
impl Upcast<dyn ParserDb> for AnalysisDatabase {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for AnalysisDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

/// Trait for converting ProjectCrate to the various crate representations
/// used by different compiler phases.
pub trait ProjectCrateExt {
    /// Convert to a semantic::Crate for semantic analysis.
    fn to_semantic_crate(&self, db: &dyn SemanticDb) -> cairo_m_compiler_semantic::Crate;
}

// Module path resolution is now centralized in cairo-m-project crate.
// This implementation delegates to the compiler's canonical module naming logic.
impl ProjectCrateExt for ProjectCrate {
    fn to_semantic_crate(&self, db: &dyn SemanticDb) -> cairo_m_compiler_semantic::Crate {
        let files = self.files(db);
        let main_module = self.main_module(db);
        let root_dir = self.root_dir(db);

        let mut modules = HashMap::new();

        // Create a temporary Project instance to use its module_name_from_path method
        let project = cairo_m_project::Project {
            manifest_path: root_dir.join("cairom.toml"),
            root_directory: root_dir.clone(),
            name: root_dir
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string(),
            source_layout: cairo_m_project::SourceLayout::default(),
            entry_point: None,
        };

        // Convert PathBuf keys to module names with proper nesting
        for (path, source_file) in files {
            // Use the Project's module_name_from_path method
            match project.module_name_from_path(&path) {
                Ok(module_name) => {
                    modules.insert(module_name, source_file);
                }
                Err(_) => {
                    // Fallback for files outside project root (shouldn't happen)
                    if let Some(module_name) = path.file_stem().and_then(|s| s.to_str()) {
                        modules.insert(module_name.to_string(), source_file);
                    }
                }
            }
        }

        cairo_m_compiler_semantic::Crate::new(
            db,
            modules,
            main_module,
            root_dir.clone(),
            root_dir
                .file_name()
                .and_then(|n| n.to_str())
                .unwrap_or("unknown")
                .to_string(),
        )
    }
}
