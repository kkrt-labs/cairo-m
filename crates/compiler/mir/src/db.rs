//! Database traits and implementations for MIR generation with Salsa integration.

use cairo_m_compiler_parser::{SourceProgram, Upcast};
use cairo_m_compiler_semantic::SemanticDb;

use crate::{MirError, MirModule};

/// Database trait for MIR (Mid-level Intermediate Representation) queries.
///
/// This trait extends SemanticDb to provide MIR-specific queries that are cached
/// and incrementally recomputed by Salsa. All MIR generation should go through
/// these queries to benefit from incremental compilation.
#[salsa::db]
pub trait MirDb: SemanticDb + Upcast<dyn SemanticDb> {}

/// Generate MIR for a source file.
///
/// This is the main entry point for MIR generation. It uses the semantic index
/// to build the MIR module, with full incremental caching support.
#[salsa::tracked]
pub fn generate_mir(db: &dyn MirDb, file: SourceProgram) -> Option<MirModule> {
    // Get the parsed module
    let parsed = cairo_m_compiler_parser::parse_program(db.upcast(), file);

    if !parsed.diagnostics.is_empty() {
        // Don't generate MIR if there are parse errors
        return None;
    }

    // Delegate to the existing generate_mir function from ir_generation
    crate::ir_generation::generate_mir(db, file).map(|arc| (*arc).clone())
}

/// Track MIR-specific errors separately for better diagnostics.
///
/// This allows us to report MIR generation errors without blocking
/// other compilation phases.
#[salsa::tracked]
pub fn mir_errors(_db: &dyn MirDb, _file: SourceProgram) -> Vec<MirError> {
    // TODO
    // For now, we'll return an empty vector
    // In the future, this should collect errors from MIR generation
    vec![]
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;

    /// Test database that implements all required traits for MIR generation
    #[salsa::db]
    #[derive(Clone, Default)]
    pub struct TestDatabase {
        storage: salsa::Storage<Self>,
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl cairo_m_compiler_parser::Db for TestDatabase {}

    #[salsa::db]
    impl SemanticDb for TestDatabase {}

    #[salsa::db]
    impl MirDb for TestDatabase {}

    impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
        fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
    }

    impl Upcast<dyn SemanticDb> for TestDatabase {
        fn upcast(&self) -> &(dyn SemanticDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
            self
        }
    }

    pub fn test_db() -> TestDatabase {
        TestDatabase::default()
    }

    #[test]
    fn test_mir_db_trait() {
        let db = TestDatabase::default();
        let source = SourceProgram::new(&db, "fn main() {}".to_string(), "test.cm".to_string());

        // This should trigger MIR generation through Salsa
        let _mir = generate_mir(&db, source);
    }
}
