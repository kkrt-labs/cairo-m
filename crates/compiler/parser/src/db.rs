#[salsa::db]
#[derive(Clone, Default)]
pub struct ParserDatabaseImpl {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for ParserDatabaseImpl {}
