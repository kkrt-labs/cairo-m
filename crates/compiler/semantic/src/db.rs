//! # Semantic Analysis Database
//!
//! This module defines the database traits and implementations for semantic analysis
//! using the Salsa incremental computation framework. It extends the parser database
//! to provide semantic-specific functionality.
//!
//! The database system enables incremental recompilation by caching query results
//! and invalidating them only when their dependencies change.

use cairo_m_compiler_diagnostics::DiagnosticCollection;
use cairo_m_compiler_parser as parser;
// We may need the parser implementation later for cross-database operations
#[allow(unused_imports)]
use parser::ParserDatabaseImpl;
use parser::{Db as ParserDb, Upcast};

/// Database trait for semantic analysis, extending the parser database
///
/// This trait defines the interface for semantic-specific database operations.
/// Type resolution and inference queries are defined as standalone tracked functions.
#[salsa::db]
pub trait SemanticDb: ParserDb + Upcast<dyn ParserDb> {
    // Type queries are defined as standalone tracked functions in type_resolution.rs
    // This trait includes any database-specific configuration or settings
}

/// Concrete database implementation for semantic analysis
///
/// This provides the actual storage and implementation for all database queries.
/// It combines both parser and semantic analysis capabilities in a single database.
///
/// # Thread Safety
///
/// This implementation is `Clone` and can be safely shared between threads.
/// Salsa handles the synchronization internally.
#[salsa::db]
#[derive(Clone, Default)]
pub struct SemanticDatabaseImpl {
    storage: salsa::Storage<Self>,
}

impl salsa::Database for SemanticDatabaseImpl {}
#[salsa::db]
impl ParserDb for SemanticDatabaseImpl {}
#[salsa::db]
impl SemanticDb for SemanticDatabaseImpl {}

impl Upcast<dyn ParserDb> for SemanticDatabaseImpl {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

/// Validate semantics of a source file and return diagnostics.
///
/// This is a tracked query that performs comprehensive semantic validation
/// including scope checking, type checking, and control flow analysis.
#[salsa::tracked]
pub fn validate_semantics(
    db: &dyn SemanticDb,
    file: parser::SourceProgram,
) -> DiagnosticCollection {
    // Parse the program first
    let parsed = parser::parse_program(db.upcast(), file);

    // Get the semantic index (this is already a tracked query)
    let index = crate::semantic_index::semantic_index_from_module(db, &parsed.module, file);

    // Create validator registry with all available validators
    let registry = crate::validation::validator::create_default_registry();
    registry.validate_all(db, file, &index)
}

#[cfg(test)]
pub(crate) mod tests {
    use super::*;

    #[salsa::db]
    #[derive(Clone)]
    pub struct TestDb {
        storage: salsa::Storage<Self>,
    }

    impl salsa::Database for TestDb {}
    #[salsa::db]
    impl ParserDb for TestDb {}
    #[salsa::db]
    impl SemanticDb for TestDb {}

    impl Upcast<dyn ParserDb> for TestDb {
        fn upcast(&self) -> &(dyn ParserDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
            self
        }
    }

    pub fn test_db() -> TestDb {
        TestDb {
            storage: salsa::Storage::default(),
        }
    }
}
