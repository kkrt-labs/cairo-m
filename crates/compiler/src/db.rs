//! Unified database implementation for the Cairo-M compiler.
//!
//! This module provides a single database that implements all compilation phases
//! (parsing, semantic analysis, MIR generation, and code generation) with full
//! Salsa incremental compilation support.

use cairo_m_compiler_codegen::CodegenDb;
use cairo_m_compiler_mir::MirDb;
use cairo_m_compiler_parser::{Db as ParserDb, Upcast};
use cairo_m_compiler_semantic::SemanticDb;

/// The main compiler database that supports all compilation phases.
///
/// This database implements all the trait requirements for:
/// - Parsing (ParserDb)
/// - Semantic analysis (SemanticDb)
/// - MIR generation (MirDb)
/// - Code generation (CodegenDb)
///
/// It provides a unified interface for the entire compilation pipeline
/// with automatic incremental compilation support through Salsa.
#[salsa::db]
#[derive(Clone, Default)]
pub struct CompilerDatabase {
    storage: salsa::Storage<Self>,
}

// Implement all required database traits
impl salsa::Database for CompilerDatabase {}

#[salsa::db]
impl ParserDb for CompilerDatabase {}

#[salsa::db]
impl SemanticDb for CompilerDatabase {}

#[salsa::db]
impl MirDb for CompilerDatabase {}

#[salsa::db]
impl CodegenDb for CompilerDatabase {}

// Implement upcast traits for each database level
impl Upcast<dyn ParserDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn ParserDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn ParserDb + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

impl Upcast<dyn MirDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn MirDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
        self
    }
}

impl Upcast<dyn CodegenDb> for CompilerDatabase {
    fn upcast(&self) -> &(dyn CodegenDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn CodegenDb + 'static) {
        self
    }
}

impl CompilerDatabase {
    /// Create a new compiler database.
    pub fn new() -> Self {
        Self::default()
    }
}
