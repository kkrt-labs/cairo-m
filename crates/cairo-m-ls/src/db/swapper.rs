use std::sync::{Arc, Mutex};
use std::time::Duration;

use tokio::sync::mpsc;
use tokio::task::JoinHandle;
use tokio::time::interval;
use tower_lsp::lsp_types::Url;
use tracing::debug;

use crate::db::AnalysisDatabase;
use crate::project::ProjectModel;
use crate::project::model::Crate;

/// Periodically swaps the analysis database to prevent unbounded memory growth.
///
/// This is a critical component for long-running language server sessions,
/// as Salsa accumulates query results over time that can lead to memory issues.
pub struct AnalysisDatabaseSwapper {
    /// Handle to the background task
    handle: Option<JoinHandle<()>>,
    /// Channel to signal shutdown
    shutdown_tx: mpsc::UnboundedSender<()>,
}

impl AnalysisDatabaseSwapper {
    /// Create a new database swapper that runs on the given interval
    pub fn new(
        db: Arc<Mutex<AnalysisDatabase>>,
        project_model: Arc<ProjectModel>,
        interval_duration: Duration,
    ) -> Self {
        let (shutdown_tx, mut shutdown_rx) = mpsc::unbounded_channel();

        let handle = tokio::spawn(async move {
            let mut timer = interval(interval_duration);
            timer.tick().await; // Skip the first immediate tick

            loop {
                tokio::select! {
                    _ = shutdown_rx.recv() => {
                        debug!("AnalysisDatabaseSwapper shutting down");
                        break;
                    }
                    _ = timer.tick() => {
                        Self::perform_swap(&db, &project_model).await;
                    }
                }
            }
        });

        Self {
            handle: Some(handle),
            shutdown_tx,
        }
    }

    /// Perform the actual database swap
    async fn perform_swap(db: &Arc<Mutex<AnalysisDatabase>>, project_model: &Arc<ProjectModel>) {
        debug!("Starting database swap");
        let start = std::time::Instant::now();

        // Step 1: Get all crates first (async call outside lock)
        let all_crates = project_model.all_crates().await;

        // Step 2: Snapshot live state with minimal lock time.
        // We get the current text of every file.
        let files_content_map = {
            let mut content_map = std::collections::HashMap::new();
            {
                let old_db = match db.lock() {
                    Ok(guard) => guard,
                    Err(_) => {
                        debug!("Failed to lock database for swap snapshot");
                        return;
                    }
                };
                for krate in &all_crates {
                    for (path, sf) in &krate.files {
                        if let Ok(uri) = Url::from_file_path(path) {
                            let content = sf.text(&*old_db).to_string();
                            content_map.insert(uri, content);
                        }
                    }
                }
                drop(old_db);
            } // Lock on old_db is released here.
            content_map
        };

        // Step 3: Build the new database and project state offline.
        // This part is computationally expensive but doesn't block the language server.
        let new_db = AnalysisDatabase::new();
        let mut new_project_crate_ids = std::collections::HashMap::new();
        let mut new_crates = std::collections::HashMap::new();

        for krate in all_crates {
            let mut new_files_in_crate = std::collections::HashMap::new();
            for path in krate.files.keys() {
                if let Ok(uri) = Url::from_file_path(path) {
                    if let Some(content) = files_content_map.get(&uri) {
                        let new_sf = cairo_m_compiler_parser::SourceFile::new(
                            &new_db,
                            content.clone(),
                            uri.to_string(),
                        );
                        new_files_in_crate.insert(path.clone(), new_sf);
                    }
                }
            }

            let main_module_name = krate
                .main_file
                .as_ref()
                .and_then(|p| p.file_stem())
                .and_then(|s| s.to_str())
                .unwrap_or("main")
                .to_string();

            let new_project_crate = crate::db::ProjectCrate::new(
                &new_db,
                krate.info.root.clone(),
                main_module_name,
                new_files_in_crate.clone(),
            );
            new_project_crate_ids.insert(krate.info.root.clone(), new_project_crate);

            // Create new Crate object with fresh SourceFile IDs
            let new_crate = Crate {
                info: krate.info.clone(),
                main_file: krate.main_file.clone(),
                files: new_files_in_crate,
            };
            new_crates.insert(krate.info.root.clone(), new_crate);
        }

        // Step 4: Update the project model first (async call)
        // CRITICAL: Must update both project_crate_ids AND crates to avoid stale Salsa IDs
        project_model
            .replace_project_crate_ids(new_project_crate_ids)
            .await;
        project_model.replace_crates(new_crates).await;

        // Step 5: Perform the atomic database swap.
        // We acquire the DB lock and swap the database.
        match db.lock() {
            Ok(mut old_db) => {
                *old_db = new_db;
                let elapsed = start.elapsed();
                debug!("Database swap completed in {:?}", elapsed);
            }
            Err(_) => {
                debug!("Failed to lock database for final swap");
            }
        }
    }

    /// Request shutdown of the swapper
    pub fn shutdown(&mut self) {
        let _ = self.shutdown_tx.send(());

        if let Some(handle) = self.handle.take() {
            handle.abort();
        }
    }
}

impl Drop for AnalysisDatabaseSwapper {
    fn drop(&mut self) {
        self.shutdown();
    }
}
