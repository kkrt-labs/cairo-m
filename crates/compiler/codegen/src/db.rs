//! Database traits and implementations for code generation with Salsa integration.

use std::sync::Arc;

use cairo_m_common::Program;
use cairo_m_compiler_mir::{MirDb, MirModule};
use cairo_m_compiler_parser::Upcast;
use cairo_m_compiler_semantic::db::Crate;

use crate::CodegenError;

/// Database trait for code generation queries.
///
/// This trait extends MirDb to provide code generation queries that are cached
/// and incrementally recomputed by Salsa. All code generation should go through
/// these queries to benefit from incremental compilation.
#[salsa::db]
pub trait CodegenDb: MirDb + Upcast<dyn MirDb> {}

/// Compile a crate to a compiled program.
///
/// This is the main entry point for code generation. It takes a crate
/// and produces the compiled program with full incremental caching support.
#[salsa::tracked]
pub fn compile_project(db: &dyn CodegenDb, crate_id: Crate) -> Result<Arc<Program>, CodegenError> {
    // Get the MIR module
    let mir_module = cairo_m_compiler_mir::db::generate_mir(db.upcast(), crate_id)
        .ok_or_else(|| CodegenError::InvalidMir("No MIR module generated".to_string()))?;

    // Use the existing compile_module logic
    let compiled = crate::compile_module(&mir_module)?;

    Ok(Arc::new(compiled))
}

/// Get the MIR module for a crate (convenience re-export).
///
/// This allows code generation to access MIR without directly depending
/// on the MIR crate's internals.
#[salsa::tracked]
pub fn codegen_mir_module(db: &dyn CodegenDb, crate_id: Crate) -> Option<Arc<MirModule>> {
    cairo_m_compiler_mir::db::generate_mir(db.upcast(), crate_id).map(Arc::new)
}

/// Track code generation errors separately for better diagnostics.
///
/// This allows us to report codegen errors without blocking other phases.
#[salsa::tracked]
pub fn codegen_errors(db: &dyn CodegenDb, crate_id: Crate) -> Vec<CodegenError> {
    // Collect errors from code generation
    match compile_project(db, crate_id) {
        Ok(_) => vec![],
        Err(e) => vec![e],
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::collections::HashMap;
    use std::path::PathBuf;

    use cairo_m_compiler_mir::MirDb;
    use cairo_m_compiler_semantic::{File, SemanticDb};

    use super::*;

    /// Test database that implements all required traits for code generation
    #[salsa::db]
    #[derive(Clone, Default)]
    pub struct TestDatabase {
        storage: salsa::Storage<Self>,
    }

    #[salsa::db]
    impl salsa::Database for TestDatabase {}

    #[salsa::db]
    impl cairo_m_compiler_parser::Db for TestDatabase {}

    #[salsa::db]
    impl SemanticDb for TestDatabase {}

    #[salsa::db]
    impl MirDb for TestDatabase {}

    #[salsa::db]
    impl CodegenDb for TestDatabase {}

    impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
        fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
            self
        }
    }

    impl Upcast<dyn SemanticDb> for TestDatabase {
        fn upcast(&self) -> &(dyn SemanticDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
            self
        }
    }

    impl Upcast<dyn MirDb> for TestDatabase {
        fn upcast(&self) -> &(dyn MirDb + 'static) {
            self
        }
        fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
            self
        }
    }

    #[test]
    fn test_codegen_db_trait() {
        let db = TestDatabase::default();
        let file = File::new(&db, "fn main() {}".to_string(), "test.cm".to_string());
        let mut modules = HashMap::new();
        modules.insert("main".to_string(), file);
        let crate_id = Crate::new(
            &db,
            modules,
            "main".to_string(),
            PathBuf::from("."),
            "test".to_string(),
        );

        // This should trigger code generation through Salsa
        let _result = compile_project(&db, crate_id);
    }
}
