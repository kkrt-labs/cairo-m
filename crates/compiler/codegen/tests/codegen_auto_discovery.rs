//! Auto-discovery test runner for CASM code generation.
//! This file automatically generates tests for all Cairo-M files in the test_data directory.

use std::path::PathBuf;

use cairo_m_compiler_codegen::{CodeGenerator, CodegenDb};
use cairo_m_compiler_mir::{generate_mir, MirDb};
use cairo_m_compiler_parser::Upcast;
use cairo_m_compiler_semantic::db::project_validate_semantics;
use cairo_m_compiler_semantic::{File, SemanticDb};

/// Test database that implements all required traits for code generation
#[salsa::db]
#[derive(Clone, Default)]
pub struct TestDatabase {
    storage: salsa::Storage<Self>,
}

#[salsa::db]
impl salsa::Database for TestDatabase {}

#[salsa::db]
impl cairo_m_compiler_parser::Db for TestDatabase {}

#[salsa::db]
impl SemanticDb for TestDatabase {}

#[salsa::db]
impl MirDb for TestDatabase {}

#[salsa::db]
impl CodegenDb for TestDatabase {}

impl Upcast<dyn cairo_m_compiler_parser::Db> for TestDatabase {
    fn upcast(&self) -> &(dyn cairo_m_compiler_parser::Db + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn cairo_m_compiler_parser::Db + 'static) {
        self
    }
}

impl Upcast<dyn SemanticDb> for TestDatabase {
    fn upcast(&self) -> &(dyn SemanticDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn SemanticDb + 'static) {
        self
    }
}

impl Upcast<dyn MirDb> for TestDatabase {
    fn upcast(&self) -> &(dyn MirDb + 'static) {
        self
    }
    fn upcast_mut(&mut self) -> &mut (dyn MirDb + 'static) {
        self
    }
}

pub fn test_db() -> TestDatabase {
    TestDatabase::default()
}

/// The result of running code generation on a test source.
pub struct CodegenOutput {
    pub casm_code: String,
    pub had_semantic_errors: bool,
}

/// Runs the full compilation pipeline from source to CASM.
pub fn check_codegen(source: &str, path: &str) -> CodegenOutput {
    use cairo_m_compiler_semantic::db::Crate;
    use std::collections::HashMap;

    let db = test_db();
    let file = File::new(&db, source.to_string(), path.to_string());

    // Create a single-file crate for MIR generation
    let mut modules = HashMap::new();
    modules.insert("main".to_string(), file);
    let crate_id = Crate::new(
        &db,
        modules,
        "main".to_string(),
        PathBuf::from("."),
        "crate_test".to_string(),
    );

    // Check for semantic errors but don't panic - we want to see what happens
    let semantic_errors = project_validate_semantics(&db, crate_id);
    let had_semantic_errors = !semantic_errors.is_empty();

    if had_semantic_errors {
        // Return early with error indication
        return CodegenOutput {
            casm_code: format!("Semantic errors found:\n{:#?}", semantic_errors),
            had_semantic_errors: true,
        };
    }

    // Generate MIR from source
    match generate_mir(&db, crate_id) {
        Ok(mir_module) => {
            let mut generator = CodeGenerator::new();
            match generator.generate_module(&mir_module) {
                Ok(_) => {
                    let casm_code = generator.debug_instructions();
                    CodegenOutput {
                        casm_code,
                        had_semantic_errors: false,
                    }
                }
                Err(e) => CodegenOutput {
                    casm_code: format!("CASM generation failed: {:?}", e),
                    had_semantic_errors: false,
                },
            }
        }
        Err(diagnostics) => CodegenOutput {
            casm_code: format!("MIR generation failed:\n{:#?}", diagnostics),
            had_semantic_errors: false,
        },
    }
}

#[test]
fn test_all_fixtures_codegen() {
    use cairo_m_test_utils::test_data_path;
    use insta::{assert_snapshot, glob, with_settings};

    // Use insta's glob! macro to discover and test all .cm files
    // This generates a separate test case for each file, allowing all snapshots
    // to be generated in a single test run
    let test_data = test_data_path();

    glob!(test_data.to_str().unwrap(), "**/*.cm", |path| {
        let source = std::fs::read_to_string(path).unwrap();

        // Extract the relative path from test_data for a cleaner name
        let relative_path = path
            .strip_prefix(&test_data)
            .unwrap_or(path)
            .to_string_lossy()
            .replace('\\', "/");

        // Generate CASM from the source code
        let codegen_output = check_codegen(&source, &relative_path);

        // Create the snapshot content
        let snapshot_content = if codegen_output.had_semantic_errors {
            format!(
                "Fixture: {}\n============================================================\nSource code:\n{}\n============================================================\nResult: SEMANTIC ERRORS\n{}\n",
                relative_path,
                source,
                codegen_output.casm_code,
            )
        } else {
            format!(
                "Fixture: {}\n============================================================\nSource code:\n{}\n============================================================\nGenerated CASM:\n{}\n",
                relative_path,
                source,
                codegen_output.casm_code,
            )
        };

        // Use with_settings to ensure consistent snapshot behavior
        with_settings!({
            description => format!("Codegen snapshot for {}", relative_path).as_str(),
            omit_expression => true
        }, {
            assert_snapshot!(snapshot_content);
        });
    });
}
