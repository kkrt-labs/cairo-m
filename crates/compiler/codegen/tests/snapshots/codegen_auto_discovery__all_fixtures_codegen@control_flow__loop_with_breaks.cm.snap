---
source: crates/compiler/codegen/tests/codegen_auto_discovery.rs
description: Codegen snapshot for control_flow/loop_with_breaks.cm
input_file: test_data/control_flow/loop_with_breaks.cm
---
Fixture: control_flow/loop_with_breaks.cm
============================================================
Source code:
fn test_loop_with_breaks() -> felt {
    let count = 0;
    let result = 0;

    loop {
        count = count + 1;

        if (count == 5) {
            continue;
        }

        result = result + count;

        if (count == 10) {
            break;
        }
    }

    return result;
}

============================================================
Generated CASM:
test_loop_with_breaks:
test_loop_with_breaks:
test_loop_with_breaks_0:
   0: 9 0 0 _              // [fp + 0] = 0
   1: 9 0 1 _              // [fp + 1] = 0
test_loop_with_breaks_1:
   2: 4 0 0 6              // [fp + 6] = [fp + 0] + 0
   3: 4 6 1 0              // [fp + 0] = [fp + 6] op 1
   4: 5 0 5 4              // [fp + 4] = [fp + 0] op 5
   5: 14 4 5 _             // if [fp + 4] != 0 jmp rel test_loop_with_breaks_4
   6: 12 9 _ _             // jump abs test_loop_with_breaks_3
test_loop_with_breaks_2:
   7: 4 1 0 -3             // Return value 0: [fp -3] = [fp + 1] + 0
   8: 11 _ _ _             // return
test_loop_with_breaks_3:
   9: 12 2 _ _             // jump abs test_loop_with_breaks_1
test_loop_with_breaks_4:
  10: 4 1 0 6              // [fp + 6] = [fp + 1] + 0
  11: 0 6 0 1              // [fp + 1] = [fp + 6] op [fp + 0]
  12: 5 0 10 5             // [fp + 5] = [fp + 0] op 10
  13: 14 5 2 _             // if [fp + 5] != 0 jmp rel test_loop_with_breaks_6
test_loop_with_breaks_5:
  14: 12 7 _ _             // jump abs test_loop_with_breaks_2
test_loop_with_breaks_6:
  15: 12 2 _ _             // jump abs test_loop_with_breaks_1
